core/tauri/src/resources/mod.rs

```diff
@@ -0,0 +1,143 @@
+/// their rid. Resources are thread-safe.
+///
+/// Resources are reference counted in Rust. This means that they can be
+/// cloned and passed around. When the last reference is dropped, the resource
+/// is automatically closed. As long as the resource exists in the resource
+/// table, the reference count is at least 1.
+pub(crate) trait Resource: Any + 'static + Send + Sync {
+  /// Returns a string representation of the resource. The default implementation
+  /// returns the Rust type name, but specific resource types may override this
+  /// trait method.
+  fn name(&self) -> Cow<'_, str> {
+    type_name::<Self>().into()
+  }
+
+  /// Resources may implement the `close()` trait method if they need to do
+  /// resource specific clean-ups, such as cancelling pending futures, after a
+  /// resource has been removed from the resource table.
+  fn close(self: Arc<Self>) {}
+}
+
+impl dyn Resource {
+  #[inline(always)]
+  fn is<T: Resource>(&self) -> bool {
+    self.type_id() == TypeId::of::<T>()
+  }
+
+  #[inline(always)]
+  pub(crate) fn downcast_arc<'a, T: Resource>(self: &'a Arc<Self>) -> Option<&'a Arc<T>> {
+    if self.is::<T>() {
+      // A resource is stored as `Arc<T>` in a BTreeMap
+      // and is safe to cast to `Arc<T>` because of the runtime
+      // check done in `self.is::<T>()`
+      let ptr = self as *const Arc<_> as *const Arc<T>;
+      Some(unsafe { &*ptr })
+    } else {
+      None
+    }
+  }
+}
+
+/// A `ResourceId` is an integer value referencing a resource. It could be
+/// considered to be the tauri equivalent of a `file descriptor` in POSIX like
+/// operating systems. Elsewhere in the code base it is commonly abbreviated
+/// to `rid`.
+pub(crate) type ResourceId = u32;
+
+/// Map-like data structure storing Tauri's resources (equivalent to file
+/// descriptors).
+///
+/// Provides basic methods for element access. A resource can be of any type.
+/// Different types of resources can be stored in the same map, and provided
+/// with a name for description.
+///
+/// Each resource is identified through a _resource ID (rid)_, which acts as
+/// the key in the map.
+#[derive(Default)]
+pub(crate) struct ResourceTable {
+  pub(crate) index: BTreeMap<ResourceId, Arc<dyn Resource>>,
+  pub(crate) next_rid: ResourceId,
+}
+
+impl ResourceTable {
+  /// Inserts resource into the resource table, which takes ownership of it.
+  ///
+  /// The resource type is erased at runtime and must be statically known
+  /// when retrieving it through `get()`.
+  ///
+  /// Returns a unique resource ID, which acts as a key for this resource.
+  pub(crate) fn add<T: Resource>(&mut self, resource: T) -> ResourceId {
+    self.add_arc(Arc::new(resource))
+  }
+
+  /// Inserts a `Arc`-wrapped resource into the resource table.
+  ///
+  /// The resource type is erased at runtime and must be statically known
+  /// when retrieving it through `get()`.
+  ///
+  /// Returns an unique resource ID, which acts as a key for this resource.
```

Comment:
idk why the github notification showed me a single change (`a` to `an`) and literally nothing else, but that's the only reason i noticed it i guess. `a unique` was correct because it starts with a `yu` sound.
And people say german is complicated... 😡 
```suggestion
  /// Returns a unique resource ID, which acts as a key for this resource.
```